ARCHITECTURE.md

```diff
@@ -0,0 +1,183 @@
+     shapes.
+ * Solids are **distinct** when they do not intersect in any way. For
+   example, their bounding boxes do not overlap. Note that this can
+   get awkward to determine, e.g. a cylinder could have a torus around
+   it whose "hole" is larger diameter than the cylinder.
+
+# Model representation
+
+The parts of the CAD we really care about are the solid parts and
+their physical properties, like material. Some operations mutate these
+parts (e.g. the imprint step) and we still want to be able to
+reference the parts consistently.
+
+CAD models tend to be hierarchical (it makes it easier to work on them
+like this) but this makes referring to parts awkward. The way I've
+chosen is to "linearise" the hierarchy of solids into a list, and
+maintain this linear order across related files and operations.
+
+For example, when checking for overlapping solids it's easy to
+determine which solids to iterate over, and when imprinting if we move
+part of one solid to another we just have to make sure we just write
+them into their appropriate indices. Material properties can be
+maintained in a separate database which just needs to be matched up
+again at the end.
+
+# OpenCascade
+
+The code uses OpenCascade, which exposes iterators (e.g.
+`TopoDS_Iterator`) that return objects in a consistent order. If this
+breaks at some point each solid could be written to its own file,
+except for the final merge step.
+
+Performance wise, OpenCascade takes ~100 times longer to open a STEP
+file than a BREP file. So if this can be done once it seems like a
+win.
+
+The OpenCascade "Pave" tool allows boolean operations to be performed
+between arbitrary solids. The tool is initialised and can then be used
+to get a union, or to cut one part out of another. Note that this API
+is awkward to use efficiently and it's easy to make the library do
+much more work than is necessary!
+
+# Tool functionality
+
+The project consists of several small tools that can be combined in a
+variety of ways to solve different tasks. These are described in the
+following sections, e.g. the CAD-CI users would just need an import,
+the overlap checker, and optionally the collector to help determine
+which solids are overlapping.
+
+# Import
+
+The source data needs to be imported and the solids linearised and
+written to a BREP file for use by other tools. As part of this these
+tools attempt to extract material properties and write them to a CSV
+file, but I don't get anything useful from the files I have at the
+moment. `step_to_brep` works with STEP files, while `brep_flatten`
+allows an existing BREP file to be worked with.
+
+`step_to_brep` contains code to validate that the input file has been
+interpreted correctly, but this might be worth splitting out into a
+separate tool as more input formats are supported.
+
+# Overlap checking
+
+Once the solids have been linearised, it's a simple matter of
+performing $N(N-1)$ checks to determine which solids intersect with
```

Comment:
What does the $N(N-1)$ syntax represent?  (It looks like some kind of Markdown formatting, but GitHub doesn't seem to interpret it as that.)

Reply:
oops, forgot about that.  Jupyter recognizes it in their notebooks as a latex style equation, forgot that github doesn't support it!